Steer/DigitizerWorkflow/src/ITSMFTDigitizerSpec.cxx

```diff
@@ -154,15 +162,48 @@ class ITSMFTDPLDigitizerTask : BaseDPLDigitizer
       mDigitizer.resetEventROFrames(); // to estimate min/max ROF for this collID
       // for each collision, loop over the constituents event and source IDs
       // (background signal merging is basically taking place here)
-      for (auto& part : eventParts[collID]) {
 
+      // first sort eventParts so that background comes last --> makes filtering out background hits based on signal easier
+      std::sort(eventParts[collID].begin(), eventParts[collID].end(), [](o2::steer::EventPart const& a, o2::steer::EventPart const& b) { return a.sourceID > b.sourceID; });
+      if (mFilterOnSignal) {
+        mHitVoxels.clear();
+      }
+      for (auto& part : eventParts[collID]) {
         // get the hits for this event and this source
-        mHits.clear();
         context->retrieveHits(mSimChains, o2::detectors::SimTraits::DETECTORBRANCHNAMES[mID][0].c_str(), part.sourceID, part.entryID, &mHits);
 
+        // fill or use voxel map for filtering
+        // This is just a demonstration based on regular 3D voxels. Other approaches seem possible, for instance to check vicinity in terms
+        // of ALIPE-chip-ID etc.;
+        if (mFilterOnSignal) {
+          if (!steer::EventPart::isBackGround(part)) {
+            for (auto& hit : mHits) {
+              auto key = mHitVoxels.getVoxelKey(vecgeom::Vector3D<float>(hit.GetX(), hit.GetY(), hit.GetZ()));
+              assert(key >= 0);
+              assert(key < 100 * 100 * 100);
+              if (!mHitVoxels.isOccupied(key)) {
+                mHitVoxels.addPropertyForKey(key, true);
+              }
+            }
+          } else {
+            // filtering step
+            if (mHitVoxels.size() > 0) {
+              auto unfiltered = mHits;
+              mHits.clear();
+              for (auto& h : unfiltered) {
+                // check region around this hit: check voxel plus few nearest neighbours ?
+                vecgeom::Vector3D<float> p(h.GetX(), h.GetY(), h.GetZ());
+                if (mHitVoxels.isOccupied(p)) {
```

Comment:
Don't you need to check also the voxels around?

Reply:
Potentially yes (as noted in the comment above). I guess this depends if a voxel covers an ALPIDE chip completely or not and how to exactly define closeness. So it's fine-tuning in some sense. Additional checks should not cost much in terms of CPU.